### PR TITLE
HPCC-29475 Add option to avoid blacklisting endpoints

### PR DESCRIPTION
**Pull request by richardkchapman**
Supported in Roxie only for now...

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [x] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [x] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [x] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

File: devdoc/roxie.md [hunk 1]
```diff
@@ -233,4 +233,51 @@ hot pages until the info is flushed, the appropriate size is not really the same
...
+   make it look like it is maintained per-activity.
+4. Is it only a failed connect that leads to blacklisting, or does a slow/error response also cause a gateway endpoint to be blacklisted? It's only a
+   failed connect.
+5. When deblacklisting, can we check any condition other than "Successfully connected"? If not, blacklisting for any reason other than "Did not connect"
+   feels like a recipe for problems. We only check for a connection (and correspondingly only blacklist for a failed connection).
+6. Are we ever using the functionality where there are more than one gateway listed? Most of the time a load-balancer is a preferable solution...
+7. The "deblacklister" thread seems to add an escalating delay between attempts. Is this delay ever reset? Is it configurable? Is it appropriate?
+8. There's a thread (in the blacklister's pool) for each blacklisted endpoint. These threads will never go away if the endpoint does not recover...
+9. There's a delay of up to 10 seconds in terminating caused by the deblacklister's connect havint to timeout before it notices that we are stopping.
```
Comment (ghalliday) [inline devdoc/roxie.md:261]: typo: having

File: devdoc/roxie.md [hunk 2]
```diff
@@ -233,4 +233,51 @@ hot pages until the info is flushed, the appropriate size is not really the same
...
+the list if the first does not respond in a timely fashion. In order to avoid this "timely fashion" check adding an overhead to every query
+made when a listed gateway is unavailable, we maintain a "blacklist" of nodes that have been seen to fail, and do not attempt to connect to
+them. There is a "deblacklister" thread that checks periodically whether it is now possible to connect to a previously-blacklisted gateway,
+and removes it from the list if so.
+
+There are a number of potential questions and issues with this code:
+1. It would appear that a blacklist is applied wven when there is only one gateway listed. In this case, the blacklist may be doing more harm than good?
+   I'm not sure that is true - it is still causing rapid failures in cases that are never going to work...
+2. Even when there is only one gateway listed, a blacklist MIGHT still be useful (you don't really want EVERY query to block trying to connect, if the
+   gateway is down - may prefer a fast failure). Also applies when there's multiple records, all being passed to a gateway, and with an ONFAIL.
```
Comment (ghalliday) [inline devdoc/roxie.md:250]: there's -> there are

File: devdoc/roxie.md [hunk 3]
```diff
@@ -233,4 +233,51 @@ hot pages until the info is flushed, the appropriate size is not really the same
...
+    Can we close the socket as well as interrupting the semaphore?
+10. Does the "reconnect" attempt from the deblacklister cause any pain for the server it is connecting to? Lots of connect attempts without any data
+    could look like a DoS attack...
+11. Retries/timeout seems to translate to
+        Owned<ISocketConnectWait> scw = nonBlockingConnect(ep, timeoutMS == WAIT_FOREVER ? 60000 : timeoutMS*(retries+1));
+    I am not sure that is correct (a single attempt to connect with a long timeout doesn't feel like it is the same as multiple attempts with shorter
+    timeouts, for example if there is a load balancer in the mix).
+11. Perhaps an option to not use blacklister would solve the immediate issue?
+12. The blacklister uses an array of endpoints - If there were a lot blacklisted, a hash table would be better
+13. Hints to control behaviour of deblacklister would behave unpredicatbly if multiple activities connected to the same endpoint with different hints
```
Comment (ghalliday) [inline devdoc/roxie.md:271]: typo: unpredictably

File: devdoc/roxie.md [hunk 4]
```diff
@@ -233,4 +233,51 @@ hot pages until the info is flushed, the appropriate size is not really the same
 We track all reads by page, and before writing also add all pages in the jhtree cache with info about the node type. Note that a hit in the
 jhtree page cache won't be noted as a read OTHER than via this last-minute add.
 
-
+## Blacklisting sockets
+
+This isn't really specific to Roxie, but was oringinally added for federated Roxie systems...
```
Comment (ghalliday) [inline devdoc/roxie.md:238]: typo: originally

**Review by ghalliday — COMMENTED** [review]
@richardkchapman a few minor comments - mainly typos

**Review by mckellyln — COMMENTED** [review]
Looks good.
If I didn't want a delay after thread tries to connect() and fails I set delay to zero, correct ?
I am not sure I like the delay += delay part after a failed connect() - if we are trying to get back to ok as soon as possible.

File: common/thorhelper/thorsoapcall.hpp [hunk 5]
```diff
@@ -27,14 +27,24 @@
...
+#define DEBLACKLIST_CONNECT_TIMEOUT 10000
+
 //Web Service Call Row Provider
 interface IWSCRowProvider : extends IInterface
 {
     virtual IHThorWebServiceCallActionArg * queryActionHelper() = 0;
     virtual IHThorWebServiceCallArg * queryCallHelper() = 0;
     virtual const void * getNextRow() = 0;
     virtual void releaseRow(const void * row) = 0;
+    virtual bool useBlacklister() const { return true; };
```
[the remark this answers is not in the document]
Comment (richardkchapman) [reply common/thorhelper/thorsoapcall.hpp:41]: It avoided changes in hThor/thor...

**Comment by richardkchapman** [comment]
> Looks good.
> If I didn't want a delay after thread tries to connect() and fails I set delay to zero, correct ?
> I am not sure I like the delay += delay part after a failed connect() - if we are trying to get back to ok as soon as possible.

I didn't like the delay += delay either - that's why I deleted it.

**Comment by mckellyln** [comment]
I might even suggest a default DEBLACKLIST_RETRY_DELAY of 0 or 10 or 100 or something very small.

**Comment by richardkchapman** [comment]
> If a soapcall to a bad url is made once, but then never made again we should eventually remove the endpoint from the list.
> Like after x minutes, or after x retry iterations ?
> Does this already happen ?

After it has retried 10 times (by default - can be changed) it will terminate the deblacklister thread and remove the endpoint from the blacklist set.

File: roxie/ccd/ccdserver.cpp [hunk 6]
```diff
@@ -27220,6 +27220,30 @@ IRoxieServerActivityFactory *createRoxieServerKeyedJoinActivityFactory(unsigned
...
+    unsigned blConnectTimeout = (unsigned) -1;
+    StringBuffer blError;
+
+public:
+    CRoxieServerSoapActivityBaseFactory(unsigned _id, unsigned _subgraphId, IQueryFactory &_queryFactory, HelperFactory *_helperFactory, ThorActivityKind _kind, IPropertyTree &_graphNode)
+        : CRoxieServerActivityFactory(_id, _subgraphId, _queryFactory, _helperFactory, _kind, _graphNode, soapStatistics)
+    {
+        if (_graphNode.getPropBool("hint[@name='noblacklist']/@value", false))
+            useBlacklist = false;
+        blDelay = _graphNode.getPropInt("hint[@name='blacklistdelay']/@value", (unsigned) -1);
```
Comment (ghalliday) [inline roxie/ccd/ccdserver.cpp:27239]: I think these values will override the defaults - so -1 will be used as the delay.

**Review by ghalliday — COMMENTED** [review]
Not ssure default values are what you would expect

File: roxie/ccd/ccdserver.cpp (hunk 6 above)
Comment (ghalliday) [reply roxie/ccd/ccdserver.cpp:27239]: The code works, but in an unexpected way,,,

**Review by ghalliday — APPROVED** [review]

**Pull request merged**